README.md
# WAF DDoS

improved-ddos-mitigation/history.txt
    1  ls
    2  git 
    3  git config --global user.name "Shivam Anand"
    4  git config --global user.email "anand.shivam44@yahoo.com"
    5  ls
    6  eval $(ssh-agent -s)
    7  ssh-add work_key 
    8  sudo chmod 400
    9  sudo chmod 400 work_key 
   10  ssh-add work_key 
   11  clear
   12  ssh-keygen -t rsa -b 4096
   13  eval $(ssh-agent -s)
   14  ssh-add id_rsa
   15  git clone -b update-readme git@github.com:anandshivam44/improved-ddos-mitigation.git
   16  cd improved-ddos-mitigation/
   17  docker
   18  docker --version
   19  sam
   20  sam --version
   21  clear
   22  docker ls -la
   23  docker ls 
   24  docker ps
   25  docker ps -a
   26  docker images
   27  sudo docker rm -vf $(docker ps -a -q)
   28  # Remove all Docker Images locally
   29  sudo docker rmi -f $(docker images -a -q)
   30  docker images
   31  sam build --use-container
   32  sam deploy --guided
   33  history
   34  sam build --use-container
   35  sam deploy
   36  sam build --use-container ; sam deploy ;
   37  sam deploy
   38  sam build --use-container ; sam deploy ;
   39  ab
   40  ab  -c 100 -n 2000 https://d4nlbe9pxnxvk.cloudfront.net/
   41  curl 2000 https://d4nlbe9pxnxvk.cloudfront.net/
   42  clear
   43  curl 2000 https://d4nlbe9pxnxvk.cloudfront.net/
   44  ab  -c 100 -n 2000 https://d4nlbe9pxnxvk.cloudfront.net/
   45  curl  https://d4nlbe9pxnxvk.cloudfront.net/
   46  curl  https://d4nlbe9pxnxvk.cloudfront.net
   47  curl ifconfig.me
   48  curl  https://d4nlbe9pxnxvk.cloudfront.net
   49  ab  -c 100 -n 2000 https://d4nlbe9pxnxvk.cloudfront.net/
   50  curl  https://d4nlbe9pxnxvk.cloudfront.net
   51  sam build --use-container ; sam deploy ;
   52  clear
   53  ab  -c 100 -n 2000 https://d4nlbe9pxnxvk.cloudfront.net/
   54  curl  https://d4nlbe9pxnxvk.cloudfront.net
   55  sam build --use-container ; sam deploy ;
   56  ab  -c 100 -n 2000 https://d4nlbe9pxnxvk.cloudfront.net/
   57  curl  https://d4nlbe9pxnxvk.cloudfront.net
   58  ab  -c 100 -n 2000 https://d4nlbe9pxnxvk.cloudfront.net/
   59  sam build --use-container ; sam deploy ;
   60  sam deploy 
   61  sam deploy --help
   62  sam deploy --no-confirm-changeset
   63  ab  -c 100 -n 2000 https://d4nlbe9pxnxvk.cloudfront.net/
   64  curl  https://d4nlbe9pxnxvk.cloudfront.net
   65  clear
   66  histry
   67  history
   68  sam build --use-container ; sam deploy --no-confirm-changeset;
   69  ab  -c 100 -n 2000 https://d4nlbe9pxnxvk.cloudfront.net/
   70  sam build --use-container ; sam deploy --no-confirm-changeset;
   71  ab  -c 100 -n 2000 https://d4nlbe9pxnxvk.cloudfront.net/
   72  curl  https://d4nlbe9pxnxvk.cloudfront.net
   73  history > history.txt


improved-ddos-mitigation/HostProcessor/host_processor.py
"""Query TimeStream DB and Update WAF Ipsets if Frequency greater than Threshold"""

import os
import datetime
import logging
import traceback
import concurrent.futures
from typing import Set, Tuple, List, Callable, Dict, Optional, TypedDict, Any, Union
import boto3

logging.basicConfig(
    format='{"logger": "%(name)s", "severity": "%(levelname)s", "line": %(lineno)d, "message": "%(message)s", "function": "%(funcName)s"}',
)

logger = logging.getLogger()
logger.setLevel(int(os.getenv("LOG_LVL", '10')))

WAF_BLOCK_IP_TABLE = os.environ['BLOCKLIST_DYNAMODB']
WAF_DDB_CONFIG_TABLE = os.environ['CONFIGURATION_DYNAMODB']

SNS_ARN = os.environ["SNS_ARN"]
THREAD_COUNT = int(os.environ['THREAD_COUNT'])

TIMESTREAM_DB_NAME = os.environ["TIMESTREAM_DB_NAME"]
TIMESTREAM_TABLE_NAME = os.environ["TIMESTREAM_TABLE_NAME"]

IPV4SET_NAME, IPV4SET_ID, _ = os.environ['IPV4SET_DETAILS'].split("|")
IPV6SET_NAME, IPV6SET_ID, _ = os.environ['IPV6SET_DETAILS'].split("|")

class IPv4Hints(TypedDict):
    ips: Set[str]
    ipset_name: str
    ipset_id: str

class IPv6Hints(TypedDict):
    ips: Set[str]
    ipset_name: str
    ipset_id: str

class IpDetails(TypedDict):
    IPv4: IPv4Hints
    IPv6: IPv6Hints

class HostDetails(TypedDict):
    duration: str
    threshold: int
    host: str
    distribution: str

# WAF V2 CloudFront is only supported in Region 'us-east-1' and Thus, cannot be changed
waf_client = boto3.client('wafv2', region_name='us-east-1')

def remove_unwanted_ip(ips: List[str]) -> List[str]:
    """Remove Private ips added while creating IPSET by cloudformation"""
    try:
        ips.remove('10.0.0.0/32')
        return ips
    except ValueError as e:
        print('error messge = ',e)
    try:
        ips.remove('fd4b:9821:be17:8c1e:0000:0000:0000:0000/128')
        return ips
    except ValueError as r:
        print('error messge r = ',r)
    return ips

def save_block_history(host_config: HostDetails, blocked_ips: List[str], ip_type: str, ipset_id: str, ipset_name: str) -> None:
    """Inserts Blocked Ip data in dynamodb table"""

    dynamodb = boto3.resource('dynamodb')
    table = dynamodb.Table(WAF_BLOCK_IP_TABLE)

    for ip in blocked_ips:
        values = {
            **host_config,
            'time': str(datetime.datetime.now()),
            'ip': ip,
            'ip_type': ip_type,
            'ipset_id': ipset_id,
            'ipset_name': ipset_name,
        }
        response = table.get_item(
            Key={
                'ip': ip,
                'distribution': host_config['distribution'],
            },
        )
        if not response.get('Item'):
            print(f"Put block ip to DDB table {WAF_BLOCK_IP_TABLE} data: {values}")
            table.put_item(
                Item=values
            )
            publish_to_sns(host_config, ip, ip_type)
    print("Added %s ip in Dynamo DB table %s", blocked_ips, WAF_BLOCK_IP_TABLE)


def update_waf_ipset(ipset_name: str, ipset_id: str, ips_to_be_blocked: Set[str], host_config: HostDetails, ip_type: str, retry: int = 3) -> Tuple[List[str], str]:
    """Add offending ips in waf ip_set"""
    print("ips_to_be_blocked 1 = ",ips_to_be_blocked)
    # ips_to_be_blocked.discard('-')
    # print("ips_to_be_blocked 2 = ",ips_to_be_blocked)
    if ip_type == "IPv4":
        ips_to_be_blocked = set(map(lambda x: x+"/32", ips_to_be_blocked))
        copy_block_ips = ips_to_be_blocked.copy()
    else:
        ips_to_be_blocked = set(map(lambda x: x+"/128", ips_to_be_blocked))
        copy_block_ips = ips_to_be_blocked.copy()

    # Get LockToken and Existing Address List
    lock_token, existing_ips = get_ipset_lock_token(waf_client, ipset_name, ipset_id)

    ips_to_be_blocked.update(existing_ips)  # Removing duplicates
    ips_to_be_blocked_list: List[str] = list(ips_to_be_blocked)

    remove_unwanted_ip(ips_to_be_blocked_list)
    print("Blocking: %s", str(ips_to_be_blocked_list))

    try:
        waf_client.update_ip_set(
            Name=ipset_name,
            Scope='CLOUDFRONT',
            Id=ipset_id,
            Addresses=ips_to_be_blocked_list,
            LockToken=lock_token
        )
    except waf_client.exceptions.WAFOptimisticLockException as e:
        logger.error(e)
        if retry > 0:
            print("Retrying Function Update Waf IpSet: %d", retry)
            return update_waf_ipset(ipset_name, ipset_id, ips_to_be_blocked, host_config, ip_type, retry=retry-1)
        else:
            logger.error("Retied %d times, IpSet not Updated", retry)
            raise Exception("Unable to Update Waf IpSet %s, retries: %d exceeded", ipset_name, retry)

    save_block_history(host_config, copy_block_ips, ip_type, ipset_id, ipset_name)
    print("Added Offending Ip Addresses to WAF IPSets")
    return copy_block_ips, ip_type

def get_ipset_lock_token(client: Callable, ipset_name: str, ipset_id: str) -> Tuple[str, List[str]]:
    """Returns WAF ip_set  lock token and ips in ipset"""

    ip_set = client.get_ip_set(
        Name=ipset_name,
        Scope='CLOUDFRONT',
        Id=ipset_id,
    )
    print(f"get_ipset json: {ip_set}")
    return ip_set['LockToken'], ip_set['IPSet']['Addresses']


def process_row(column_info: List[str], row: Dict[str, Any], threshold: int) -> Tuple[Optional[Union[str, List[str]]], Optional[str]]:
    '''if 'no of request' > 'threshold' then return the ip and type of ip'''
    try:
        row_dict = {column["Name"]:value["ScalarValue"] for column, value in zip(column_info, row["Data"])}

        if int(row_dict["Frequency"]) > threshold:
            print("Row debug: %s", str(row_dict))

            if row_dict.get('c_ip'):
                return row_dict["c_ip"], row_dict["c_ip_version"]
            elif row_dict.get('x_forwarded_for'):
                return [i.strip() for i in row_dict["x_forwarded_for"].split(",")], row_dict["c_ip_version"]
            else:
                raise Exception(f"No ip found, frequency > threshold, Row: {row_dict}")
        else:
            return None, None
    except Exception as e:
        logger.exception(f"Error in process_row function: {e}")
        return None, None

def process_host(header: str, host: str, duration: str, threshold: int, offending_ips: IpDetails) -> IpDetails:
    """Query Timestream db for logs and return ips where count greater than threshold"""

    query_client = boto3.client('timestream-query')
    paginator = query_client.get_paginator('query')

    query = f'''SELECT {header}, cs_host, cs_uri_stem, x_host_header, c_ip_version, COUNT({header}) AS Frequency FROM "{TIMESTREAM_DB_NAME}"."{TIMESTREAM_TABLE_NAME}"  WHERE cs_host='{host}' AND time between ago({duration}) and now() AND {header} != '-' GROUP BY {header},cs_host,cs_uri_stem,c_ip_version,x_host_header ORDER BY Frequency DESC '''
    print("Query String = %s", query)

    try:
        page_iterator = paginator.paginate(QueryString=query)

        for page in page_iterator:
            print("page = ",str(page))
            column_info = page['ColumnInfo']
            for row in page['Rows']:
                print("Row = ", str(row))
                ip, version = process_row(column_info, row, threshold)
                if isinstance(ip, str):
                    offending_ips[version]["ips"].add(ip)
                elif isinstance(ip, list):
                    offending_ips[version]["ips"].update(ip)

    except Exception as err:
        logger.exception("ERROR IN TIMESTREAM QUERY") # Column_info: {column_info}, Row: {row} Threshold: {threshold}, Host: {host}")
        traceback.print_exc()
        raise Exception(f"Exception while running query: {err}")
    else:
        # print("TimeStream Query Success")
        return offending_ips

# def publish_to_sns(host: str, distribution: str, threshold: int, duration: str, offending_ips: IpDetails) -> None:
def publish_to_sns(host_config: HostDetails, ip: str, ip_version: str) -> None:
    """Publish Blocked ip details to sns"""

    subject = f"Successfully Blocked a IP"
    message = f"""
        ------------------------------------------
        IP Blocked by WAF
        ------------------------------------------
        {ip_version:<20}    :{ip}
        {'Host':<20}        :{host_config['host']}
        {'Distribution':<20}:{host_config['distribution']}
        {'Threshold':<20}   :{host_config['threshold']}
        {'Duration':<20}    :{host_config['duration']}
        ------------------------------------------
        """

    sns = boto3.client("sns")
    response = sns.publish(
        TopicArn=SNS_ARN,
        Message=message,
        Subject=subject,
    )

    print("Published to SNS: %s", str(response))

def get_config(host: str, distribution: str) -> Tuple[str, int]:
    """This function reads configuration data from dynamodb table"""

    dynamodb = boto3.resource('dynamodb')
    table = dynamodb.Table(WAF_DDB_CONFIG_TABLE)

    response = table.get_item(
        Key={
            "distribution": distribution,
            "host": host
        }
    )
    print(f"Configurations details from Dynamo db: {response}")
    return response['Item']['duration'], int(response['Item']['threshold'])

def check_ip(ips: Set[str], ip_version: str) -> bool:
    """check offending ip is found or not"""
    if ips:
        print("Found %d new %s Addresses", len(ips), ip_version)
        return True
    else:
        print("No %s Addresses found", ip_version)
        return False

def lambda_handler(event: str, context) -> None:
    """Main Fn: Adds offending ips to WAf Block List based on threshold"""

    print("event = ",event)

    retry: int = 3
    offending_ips_info: IpDetails = {
        "IPv4": {
            "ips": set(),
            "ipset_name": IPV4SET_NAME,
            "ipset_id": IPV4SET_ID,
        },
        "IPv6": {
            "ips": set(),
            "ipset_name": IPV6SET_NAME,
            "ipset_id": IPV6SET_ID,
        }
    }

    host, distribution = event['host_details'].split(',')

    duration, threshold = get_config(host, distribution)
    host_config: HostDetails = {
        "duration": duration,
        "threshold": threshold,
        "host": host,
        "distribution": distribution,
    }
    print(f"{host_config} from table {WAF_DDB_CONFIG_TABLE}")

    headers = ("x_forwarded_for", "c_ip")
    with concurrent.futures.ThreadPoolExecutor(max_workers=THREAD_COUNT) as executor:
        future_to_project = {executor.submit(process_host, header, host, duration, threshold, offending_ips_info): header for header in headers}
        for future in concurrent.futures.as_completed(future_to_project):
            processed = future_to_project[future]
            try:
                data = future.result()
            except Exception as exc:  # pylint: disable=broad-except
                logger.error("%r generated an exception: %s", processed, exc)
                traceback.print_exc()
            else:
                print("%s query status  %s", processed, data)

    print(f"offending_ips_info: {offending_ips_info}")

    with concurrent.futures.ThreadPoolExecutor(max_workers=THREAD_COUNT) as executor:
        future_to_project = {executor.submit(update_waf_ipset, ip_details["ipset_name"], ip_details["ipset_id"], ip_details["ips"], host_config, ip_version, retry): (ip_version, ip_details) for ip_version, ip_details in offending_ips_info.items() if check_ip(ip_details["ips"], ip_version)}
        for future in concurrent.futures.as_completed(future_to_project):
            processed = future_to_project[future]
            try:
                data = future.result()
            except Exception as exc: # pylint: disable=broad-except
                logger.error("%r generated an exception: %s", processed, exc)
                traceback.print_exc()
            else:
                print("%r status  %s", processed, data)
                # print("Updated IPSet %s with %d IP's", ip_details["ipset_name"], len(data))
                print('Blocked Ips: %s', str(data))




QueryTimeStream/Query.py
# import json
import boto3
import time
# boto3.setup_default_session(profile_name='ssdev-1')
print("Program started")

ipset_name='test-ip-set-name'
ipset_id='5511edd2-e521-4a31-a20b-7cac41be38e1'
 
def isIPv4(IP):
    try:
        return True if type(ip_address(IP)) is IPv4Address else False
    except ValueError:
        return "Invalid"

def update_waf_ipset(ipset_name,ipset_id,ip_to_be_blocked):
    ip_to_be_blocked=ip_to_be_blocked+"/32"
    """Updates the AWS WAF IP set"""
    session = boto3.session.Session(profile_name='ssdev-1')
    waf_client = session.client('wafv2',region_name='us-east-1')
    
    # waf_client = boto3.client('wafv2',region_name='us-east-1')

    lock_token, address_list = get_ipset_lock_token(waf_client,ipset_name,ipset_id)
    address_list.append(ip_to_be_blocked)
    waf_client.update_ip_set(
        Name=ipset_name,
        Scope='CLOUDFRONT',
        Id=ipset_id,
        Addresses=address_list,
        LockToken=lock_token
    )

    print(f'Updated IPSet "{ipset_name}" with {len(address_list)} CIDRs')

def get_ipset_lock_token(client,ipset_name,ipset_id):
    """Returns the AWS WAF IP set lock token"""
    ip_set = client.get_ip_set(
        Name=ipset_name,
        Scope='CLOUDFRONT',
        Id=ipset_id)
    print("Inside get_ipset_lock_token")
    print(ip_set)
    # print(type(ip_set['IPSet']['Addresses']))
    
    return ip_set['LockToken'],ip_set['IPSet']['Addresses']
    


def query():
    # TODO implement
    # print("QUERY STARTED")
    # boto3.setup_default_session(profile_name='ssdev-1')
    session = boto3.session.Session(profile_name='ssdev-1')
    query_client = session.client('timestream-query')
    paginator = query_client.get_paginator('query')
    
    DATABASE_NAME='CloudFrontLogsTimeSeriesDb-xBECBoapfI2U'
    TABLE_NAME='RealtimeLogsTable-7MYNnsSemTCE'
    DURATION='15s'
    
    # See records ingested into this table so far
    # QUERY = f''''SELECT  c_ip,cs_host,cs_uri_stem,x_host_header, COUNT(c_ip) AS Frequency FROM "{DATABASE_NAME}"."{TABLE_NAME}" WHERE time between ago(15m) and now() GROUP BY c_ip,cs_host,cs_uri_stem,x_host_header ORDER BY Frequency DESC '''
    QUERY = f'''SELECT c_ip, cs_host, cs_uri_stem, x_host_header, COUNT(c_ip) AS Frequency FROM "{DATABASE_NAME}"."{TABLE_NAME}"  WHERE time between ago({DURATION}) and now() GROUP BY c_ip,cs_host,cs_uri_stem,x_host_header ORDER BY Frequency DESC '''
    # QUERY='SELECT * FROM "CloudFrontLogsTimeSeriesDb-xBECBoapfI2U"."RealtimeLogsTable-7MYNnsSemTCE" WHERE time between ago(150m) and now() ORDER BY time DESC LIMIT 10 '
    
    try:
        page_iterator = paginator.paginate(QueryString=QUERY)
        for page in page_iterator:
            column_info = page['ColumnInfo']
            for row in page['Rows']:
                _parse_row(column_info, row)
    except Exception as err:
        print("THERE WAS AN ERROR")
        print("Exception while running query:", err)
    # page_iterator = paginator.paginate(QueryString=QUERY)
    # for page in page_iterator:
    #     column_info = page['ColumnInfo']
    #     for row in page['Rows']:
    #         _parse_row(column_info, row)

    
def _parse_row(column_info, row):
    data = row['Data']
    # row_output = []
    # for j in range(len(data)):
    #     info = column_info[j]
        # datum = data[j]
    #     print(info)
    
    print(data[0]['ScalarValue'],' ',data[4]['ScalarValue'])    
    if int(data[4]['ScalarValue'])>30:
        ip_to_be_blocked=data[0]['ScalarValue']
        print('BLOCKing',ip_to_be_blocked,'because it made',data[4]['ScalarValue'],'attempts')
        # if isIPv4(ip_to_be_blocked):
        #     address_list=get_address_list_waf_ipset(ipset_name,ipset_id)
        #     address_list=['10.0.0.2/32']
        #     update_waf_ipset(ipset_name,ipset_id,address_list)
        update_waf_ipset(ipset_name,ipset_id,ip_to_be_blocked)
        # row_output.append(self._parse_datum(info, datum))

        # return "{%s}" % str(row_output)
def repeat():
    while True:
        print ("tick")
        query()
        time.sleep(2)
repeat()
print("Program Ended")